### PR TITLE
feat: Add Segment tracking for Product List Viewed and Product Clicked events

### PR DESCRIPTION
**Pull request by mikemurray**
Issue: #56
Impact: **minor**
Type: **feat**

## Issue

Add event tracking for `Product List Viewed` and `Product Clicked` event for the homepage and tag pages.

## Solution

- Add docs outlining data mappings
- Update `lib/tracking/trackProductListViewed.js` to include outlined data mappings
- Add HOC for tracking product clicked event
- Ensure the `Product List Viewed` is triggered when viewing the homepage and tag pages

## Testing

1. Create a segment account and get the API key. Add the key to the `.env` (see `.env.example`)
2. Start the app navigate to the Product grid homepage and tag pages
3. In the segment dashboard under `sources > javascript > debugger` see the event being tracked with proper data mappings

You should see a `Product List Viewed`, and `Product Clicked` events  something like the following examples:

- https://segment.com/docs/spec/ecommerce/v2/#product-list-viewed 
- https://segment.com/docs/spec/ecommerce/v2/#product-clicked

__Note: The mappings are not going to be 100% the same as segments suggestions.__

### CODE REVIEW

File: docs/tracking-events.md [hunk 1]
```diff
@@ -250,8 +254,109 @@ class Page extends Component {
...
+
+Data for the Segment e-commerce event [Product List Viewed](https://segment.com/docs/spec/ecommerce/v2/#product-list-viewed), will require values from a `CatalogItemProduct` and `CatalogItemVariant`.
+
+```js
+{
+  // ID of the list being viewed. Should be a tag id or home if it's the homepage
+  list_id: "",
+
+  // Name of tag or Home
+  category: "",
```
Comment (kieckhafer) [inline docs/tracking-events.md]: `Product list viewed` does not seem to be getting the `category` data, it's always blank, no matter which list I'm on.

File: src/lib/tracking/trackProductListViewed.js [hunk 2]
```diff
@@ -1,41 +1,18 @@
...
  */
-export default (options) => (Component) => {
-  const component = class trackProductListViewed extends React.Component {
-    render() {
-      return (
-        <Component {...this.props} />
-      );
-    }
+export default (options) => track(({ tag, catalogItems }) => {
+  const products = (Array.isArray(catalogItems) && catalogItems.map((catalogItem) => catalogItem.node.product)) || undefined;
```
Comment (kieckhafer) [inline src/lib/tracking/trackProductListViewed.js]: Do we need a protection here to check for `node`, similar to the other Segment PR? We might not, I haven't had any errors yet.

Comment (mikemurray) [reply src/lib/tracking/trackProductListViewed.js]: I don't think so, because node should exist if there are items in the array.

**Comment by mikemurray** [comment]
@kieckhafer The category fields just take the first tag from the product. Since we have no real weight mechanic, I figured the first tag must be the most imported, regardless of the page it's on.

So, even a tag page, the category will still be the first tag for that product. The `list_id` is better served to identify what page these related products belong to.

File: docs/tracking-events.md (hunk 1 above)
Comment (mikemurray) [reply docs/tracking-events.md]: @kieckhafer I'm getting the category when I visit a tag page

screen shot 2018-06-12 at 11 01 32 am

**Comment by mikemurray** [comment]
@kieckhafer I'm not sure, but the multiple renders may be related to `withTag` and `withCatalogItems`. The component renders exactly twice, and my assumption is that both are doing a graphql query and that makes the component render twice.

**Comment by mikemurray** [comment]
@kieckhafer added a ticket about the multiple re-renders / multiple tracking events dispatched. https://github.com/reactioncommerce/reaction-next-starterkit/issues/142

**Comment by kieckhafer** [comment]
Looks good with the changes, and the new ticket added to fix the issue.

👍

**Pull request merged**